test/e2e/activator_test.go

```diff
@@ -0,0 +1,128 @@
+func TestActivatorOverload(t *testing.T) {
+	t.Parallel()
+	const (
+		// The number of concurrent requests to hit the activator with.
+		// 1000 = the number concurrent connections in Istio.
+		concurrency = 1000
+		// Timeout to wait for the responses.
+		// Ideally we must wait ~30 seconds, TODO: need to figure out where the delta comes from.
+		timeout = 65 * time.Second
+		// How long the service will process the request in ms.
+		serviceSleep = 300
+	)
+	var (
+		clients    = Setup(t)
+		group      errgroup.Group
+		resChannel = make(chan *spoof.Response, concurrency)
+	)
+
+	names := test.ResourceNames{
+		Service: test.ObjectNameForTest(t),
+		Image:   "observed-concurrency",
+	}
+
+	fopt := func(service *v1alpha1.Service) {
+		service.Spec.RunLatest.Configuration.RevisionTemplate.Spec.ContainerConcurrency = 1
+		service.Spec.RunLatest.Configuration.RevisionTemplate.Annotations = map[string]string{"autoscaling.knative.dev/maxScale": "10"}
+	}
+
+	test.CleanupOnInterrupt(func() { test.TearDown(clients, names) })
+	defer test.TearDown(clients, names)
+
+	t.Logf("Creating a service with run latest configuration.")
+	// Create a service with concurrency 1 that could sleep for N ms.
+	// Limit its maxScale to 10 containers, wait for the service to scale down and hit it with concurrent requests.
+	resources, err := test.CreateRunLatestServiceReady(t, clients, &names, &test.Options{}, fopt)
+	if err != nil {
+		t.Fatalf("Unable to create resources: %v", err)
+	}
+	domain := resources.Route.Status.Domain
+
+	deploymentName := resources.Revision.Name + "-deployment"
+
+	t.Logf("Waiting for deployment to scale to zero.")
+	if err := pkgTest.WaitForDeploymentState(
+		clients.KubeClient,
+		deploymentName,
+		test.DeploymentScaledToZeroFunc,
+		"DeploymentScaledToZero",
+		test.ServingNamespace,
+		3*time.Minute); err != nil {
+		t.Fatalf("Failed waiting for deployment to scale to zero: %v", err)
+	}
+
+	url := fmt.Sprintf("http://%s/?timeout=%d", domain, serviceSleep)
+
+	client, err := pkgTest.NewSpoofingClient(clients.KubeClient, t.Logf, domain, test.ServingFlags.ResolvableDomain)
+	client.RequestTimeout = timeout
+
+	t.Logf("Starting to send out the requests")
+
+	// Send requests async and wait for the responses.
+	for i := 0; i < concurrency; i++ {
+		group.Go(func() error {
+			req, err := http.NewRequest(http.MethodGet, url, nil)
+			if err != nil {
+				return fmt.Errorf("error creating http request: %v", err)
+			}
+			res, err := client.Do(req)
+			if err != nil {
+				return fmt.Errorf("unexpected error sending a request, %v", err)
+			}
+			resChannel <- res
+			return nil
+		})
+	}
+	if err := group.Wait(); err != nil {
+		t.Fatalf("unexpected error making requests against activator: %v", err)
+	}
+	t.Logf("Done sending out requests")
+	close(resChannel)
+
+	t.Logf("Process the responses")
+
+	for resp := range resChannel {
+		if resp != nil {
```

Comment:
Can this actually be `nil`? Considering that you don't put `nil` in the channel, when request fails (the point where `client.Do` returns an error).